Python/parking_lot.c

```diff
@@ -0,0 +1,470 @@
+    pthread_mutex_unlock(&sema->mutex);
+#endif
+    return res;
+}
+
+int
+_PySemaphore_Wait(_PySemaphore *sema, _PyTime_t timeout, int detach)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    int was_attached = 0;
+    if (tstate) {
+        was_attached = (tstate->_status.active);
+        if (was_attached && detach) {
+            PyEval_ReleaseThread(tstate);
+        }
+    }
+
+    int res = _PySemaphore_PlatformWait(sema, timeout);
+
+    if (tstate) {
+        if (was_attached && detach) {
+            PyEval_AcquireThread(tstate);
+        }
+    }
+    return res;
+}
+
+void
+_PySemaphore_Wakeup(_PySemaphore *sema)
+{
+#if defined(MS_WINDOWS)
+    if (!ReleaseSemaphore(sema->platform_sem, 1, NULL)) {
+        Py_FatalError("parking_lot: ReleaseSemaphore failed");
+    }
+#elif defined(USE_SEMAPHORES)
+    int err = sem_post(&sema->platform_sem);
+    if (err != 0) {
+        Py_FatalError("parking_lot: sem_post failed");
+    }
+#else
+    pthread_mutex_lock(&sema->mutex);
+    sema->counter++;
+    pthread_cond_signal(&sema->cond);
+    pthread_mutex_unlock(&sema->mutex);
+#endif
+}
+
+_PySemaphore *
+_PySemaphore_Alloc(void)
+{
+    // Make sure we have a valid thread_data. We need to acquire
+    // some locks before we have a fully initialized PyThreadState.
+    _PyParkingLot_InitThread();
+
+    ThreadData *this_thread = thread_data;
+    if (this_thread->depth >= MAX_SEMA_DEPTH) {
+        Py_FatalError("_PySemaphore_Alloc(): too many calls");
+    }
+    return &this_thread->semas[this_thread->depth++];
+}
+
+void
+_PySemaphore_Free(_PySemaphore *sema)
+{
+    ThreadData *this_thread = thread_data;
+    this_thread->depth--;
+    if (&this_thread->semas[this_thread->depth] != sema) {
+        Py_FatalError("_PySemaphore_Free(): mismatch wakeup");
+    }
+    _PyParkingLot_DeinitThread();
+}
+
+void
+_PyParkingLot_InitThread(void)
+{
+    if (thread_data != NULL) {
+        thread_data->refcount++;
+        return;
+    }
+    ThreadData *this_thread = PyMem_RawMalloc(sizeof(ThreadData));
+    if (this_thread == NULL) {
+        Py_FatalError("_PyParkingLot_InitThread: unable to allocate thread data");
+    }
+    memset(this_thread, 0, sizeof(*this_thread));
+    this_thread->refcount = 1;
+    this_thread->depth = 0;
+    for (int i = 0; i < MAX_SEMA_DEPTH; i++) {
+        _PySemaphore_Init(&this_thread->semas[i]);
+    }
+    thread_data = this_thread;
+}
+
+void
+_PyParkingLot_DeinitThread(void)
+{
+    ThreadData *td = thread_data;
+    if (td == NULL) {
+        return;
+    }
+
+    if (--td->refcount != 0) {
+        assert(td->refcount > 0);
+        return;
+    }
+
+    thread_data = NULL;
+    for (int i = 0; i < MAX_SEMA_DEPTH; i++) {
+        _PySemaphore_Destroy(&td->semas[i]);
+    }
+
+    PyMem_RawFree(td);
+}
+
+static void
+enqueue(Bucket *bucket, const void *address, struct wait_entry *wait)
+{
+    if (!bucket->root.next) {
+        // initialize bucket
+        llist_init(&bucket->root);
+    }
```

Comment:
FWIW, it would be nice if the linked list didn't leak implementation details like this.  We can be explicit about it:

```suggestion
    if (!llist_is_initialized(&bucket->root)) {
        // initialize bucket
        llist_init(&bucket->root);
    }
```

Otherwise, I'd expect `llist_init()` to do the right thing here (like we do with `_PyParkingLot_InitThread()`:

```suggestion
    // initialize bucket
    llist_init(&bucket->root);
```

or even

```suggestion
    // initialize bucket
    llist_maybe_init(&bucket->root);
```

Reply:
I'd like avoid assigning meaning to `NULL` pointers in the linked-list abstractions, since the meaning is dependent on the context. I've changed the code to statically initialize buckets properly, which avoids the need to lazy-initialize in `enqueue`.